directory/utils.bzl

```diff
@@ -0,0 +1,107 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Skylib module containing utility functions related to directories."""
+
+visibility("public")
+
+_NOT_FOUND = """{directory} does not contain an entry named {name}.
+Instead, it contains the following entries:
+{children}
+
+"""
+_WRONG_TYPE = "Expected {dir}/{name} to have type {want}, but got {got}"
+
+def _check_path_relative(path):
+    if path.startswith("/"):
+        fail("Path must be relative. Got {path}".format(path = path))
+
+def get_children(directory):
+    """Gets all children of a directory.
+
+    Args:
+        directory: The directory to look within.
+
+    Returns:
+        dict[str, File|DirectoryInfo]
+    """
+    return {
+        name: getattr(directory.entries, name).value
+        for name in sorted(dir(directory.entries))
+    }
+
+def get_child(directory, name, require_dir = False, require_file = False):
+    """Gets the direct child of a directory.
+
+    Args:
+        directory: (DirectoryInfo) The directory to look within.
+        name: (string) The name of the directory/file to look for.
+        require_dir: (bool) If true, throws an error if the value is not a
+            directory.
+        require_file: (bool) If true, throws an error if the value is not a
+            file.
+
+    Returns:
+        (File|DirectoryInfo) The content contained within.
+    """
+    entry = getattr(directory.entries, name, None)
+    if entry == None:
+        fail(_NOT_FOUND.format(
+            directory = directory.human_readable,
+            name = repr(name),
+            children = "\n".join(get_children(directory).keys()),
+        ))
+    result = entry.value
+    if require_dir and type(result) == "File":
```

Comment:
Nit: Should we test for `!= "Directory"` instead? I'm not sure if we'd ever have other types (Symlink?), but if so it's cleaner to not assume that "Directory" == not "File".

Reply:
`type(DirectoryInfo(...))` returns `struct`. I felt that checking that it wasn't a file was better than checking that it was a struct.